CommonTools/ParticleFlow/plugins/PFCandidateRecalibrator.cc

```diff
@@ -0,0 +1,326 @@
+#include <iostream>
+
+
+struct HEChannel { 
+  float eta; float phi; float ratio;
+  HEChannel(float eta, float phi, float ratio): 
+    eta(eta),
+    phi(phi),
+    ratio(ratio)
+  {}
+};
+struct HFChannel { 
+  int ieta; int iphi; int depth; float ratio;
+  HFChannel(int ieta, int iphi, int depth, float ratio): 
+    ieta(ieta),
+    iphi(iphi),
+    depth(depth),
+    ratio(ratio)
+  {}  
+};
+
+
+class PFCandidateRecalibrator : public edm::stream::EDProducer<> {
+    public:
+        PFCandidateRecalibrator(const edm::ParameterSet&);
+        ~PFCandidateRecalibrator() override {};
+
+    private:
+        void beginRun(const edm::Run& iRun, edm::EventSetup const& iSetup) override;
+        void endRun(const edm::Run& iRun, edm::EventSetup const& iSetup) override;
+        void produce(edm::Event&, const edm::EventSetup&) override;
+
+        edm::ESWatcher<HcalRecNumberingRecord> hcalDbWatcher_;
+        edm::ESWatcher<HcalRespCorrsRcd> hcalRCWatcher_;
+  
+        edm::EDGetTokenT<reco::PFCandidateCollection> pfcandidates_;
+
+        std::vector<HEChannel> badChHE_;
+        std::vector<HFChannel> badChHF_;
+};
+
+PFCandidateRecalibrator::PFCandidateRecalibrator(const edm::ParameterSet &iConfig) :
+  pfcandidates_(consumes<reco::PFCandidateCollection>(iConfig.getParameter<edm::InputTag>("pfcandidates")))
+{
+    produces<reco::PFCandidateCollection>();
+    produces<reco::PFCandidateCollection>("discarded");
+    produces<edm::ValueMap<reco::PFCandidateRef>>();
+}
+
+void PFCandidateRecalibrator::beginRun(const edm::Run &iRun, const edm::EventSetup &iSetup)
+{
+    if (hcalDbWatcher_.check(iSetup) || hcalRCWatcher_.check(iSetup))
+      {
+	//Get Calib Constants from current GT
+	edm::ESHandle<HcalDbService> gtCond;
+	iSetup.get<HcalDbRecord>().get(gtCond);
+	
+	//Get Calib Constants from bugged tag
+	edm::ESHandle<HcalTopology> htopo;
+	iSetup.get<HcalRecNumberingRecord>().get(htopo);
+	const HcalTopology* theHBHETopology = htopo.product();
+	
+	edm::ESHandle<HcalRespCorrs> buggedCond;
+	iSetup.get<HcalRespCorrsRcd>().get("bugged", buggedCond);
+	HcalRespCorrs buggedRespCorrs(*buggedCond.product());
+	buggedRespCorrs.setTopo(theHBHETopology);
+	
+	//access calogeometry
+	edm::ESHandle<CaloGeometry> calogeom;
+	iSetup.get<CaloGeometryRecord>().get(calogeom);
+	const CaloGeometry* cgeo = calogeom.product();
+	const HcalGeometry* hgeom = static_cast<const HcalGeometry*>(cgeo->getSubdetectorGeometry(DetId::Hcal,HcalForward));
+
+	//reset the bad channel containers
+	badChHE_.clear();
+	badChHF_.clear();
+	
+	//fill bad cells HE (use eta, phi)
+	const std::vector<DetId>& cellsHE = hgeom->getValidDetIds(DetId::Detector::Hcal, HcalEndcap);
+	for( auto id : cellsHE)
+	  {
+	    float currentRespCorr = gtCond->getHcalRespCorr(id)->getValue();
+	    float buggedRespCorr = buggedRespCorrs.getValues(id)->getValue();
+	    float ratio = currentRespCorr/buggedRespCorr;
+	    
+	    if(ratio != 1.f)
```

Comment:
repeating https://github.com/cms-sw/cmssw/pull/22941#discussion_r182114731

is this not too strict?
I worry about possible round-offs in filling the two tables. So, maybe a delta of 0.001 from 1 is a better check.

Reply:
agreed, will be there at the next commit